src/System.Net.HttpListener/tests/HttpListenerAuthenticationTest.cs

```diff
@@ -384,22 +390,21 @@ public void Realm_SetDisposed_ThrowsObjectDisposedException()
 
         public async Task<HttpResponseMessage> AuthenticationFailure(HttpClient client, HttpStatusCode errorCode)
         {
+            client.Timeout = new TimeSpan(0, 0, 1);
             Task<HttpResponseMessage> clientTask = client.GetAsync(_factory.ListeningUrl);
 
             // The server task will hang forever if it is not cancelled.
             var tokenSource = new CancellationTokenSource();
             Task<HttpListenerContext> serverTask = Task.Run(() => _listener.GetContext(), tokenSource.Token);
-
+            
             // The client task should complete first - the server should send a 401 response.
             Task resultTask = await Task.WhenAny(clientTask, serverTask);
             tokenSource.Cancel();
             if (resultTask == serverTask)
             {
-                await serverTask;
+                await clientTask;
```

Comment:
>  the change you made is very likely going to result in deadlock 

I don't understand the deadlock. The client has a default timeout of 100s (or, with my change of 1s) so even if the server completes first with an error, the client will timeout eventually.

> And why do we care about compiler optimizations in a test?

I believe that the test code is representing our functional spec as well as samples for people that want to write their own applications.

>  And it's perfectly fine to await something that's already finished.

Is this the recommend pattern for our customers? 

> Every time any Assert.whatever fails it's something that's "very wrong", and an "Assert.whatever" failing in a test just generates an exception. This is simply propagating that failure out to the calling test. I don't understand what the objection here is.

My objection is that we need to crash fast and get all information under the constraints of the current test system. The Assert method you present above will give us virtually no information about the side that's hanging.